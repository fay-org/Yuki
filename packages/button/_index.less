.fect-button {
  box-sizing: border-box;
  display: inline-block;
  font-weight: 400;
  user-select: none;
  text-transform: capitalize;
  white-space: nowrap;
  text-align: center;
  outline: none;
  width: auto;
  height: 40px;
  border-radius: 5px;
  font-size: 15px;
  color: var(--accents-5);
  padding: 0 25px;
  overflow: hidden;
  cursor: pointer;
  position: relative;
  transition: border 0.2s, background 0.2s, color 0.2s ease-out;
  &--success {
    border-color: var(--success-default);
    background-color: var(--success-default);
    color: #fff;
  }
  &--warning {
    border-color: var(--warning-default);
    background-color: var(--warning-default);
    color: #fff;
  }
  &--error {
    background-color: var(--error-default);
    border-color: var(--error-default);
    color: #fff;
  }
}
