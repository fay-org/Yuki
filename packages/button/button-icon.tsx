import { defineComponent, PropType } from 'vue'

const ButtonIcon = defineComponent({
  props: {
    icon: String as PropType<keyof HTMLElementTagNameMap>,
  },
  setup(props) {
    return () => <span class="fect-button__icon">{props.icon}</span>
  },
})

export default ButtonIcon
