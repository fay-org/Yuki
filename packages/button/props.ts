import { PropType } from 'vue'
import {
  NormalSizes,
  ButtonTypes,
  LoadingTypes,
} from '../utils/theme/propTypes'
import { UnknowProp } from '../utils/base'

export const props = {
  type: {
    type: String as PropType<ButtonTypes>,
    default: 'default',
  },
  size: {
    type: String as PropType<NormalSizes>,
    default: 'medium',
  },
  ghost: Boolean,
  loading: Boolean,
  shadow: Boolean,
  auto: Boolean,
  disabled: Boolean,
  icon: UnknowProp,
  loadType: {
    type: String as PropType<LoadingTypes>,
    default: 'deafult',
  },
  effect: Boolean,
}
