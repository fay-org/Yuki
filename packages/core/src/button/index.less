@keyframes button-expand {
  0% {
    opacity: 0;
    transform: scale(1);
  }

  30% {
    opacity: 1;
  }

  80% {
    opacity: 0.5;
  }

  100% {
    transform: scale(28);
    opacity: 0;
  }
}

.fect-button {
  box-sizing: border-box;
  display: inline-block;
  line-height: var(--button-height);
  border-radius: var(--button-radius);
  font-weight: 400;
  font-size: var(--button-font-size);
  user-select: none;
  outline: none;
  text-transform: capitalize;
  justify-content: center;
  text-align: center;
  white-space: nowrap;
  transition: background-color 200ms ease 0ms, box-shadow 200ms ease 0ms, border 200ms ease 0ms, color 200ms ease 0ms;
  position: relative;
  overflow: hidden;
  cursor: pointer;
  height: var(--button-height);
  min-width: var(--button-min-width);
  width: initial;
  padding: var(--button-pt) var(--button-pr) var(--button-pb) var(--button-pl);
  margin: var(--button-mt) var(--button-mr) var(--button-mb) var(--button-ml);
  background-color: var(--button-bg-color);
  border: 1px solid var(--button-border-color);
  color: var(--button-color);

  &--disabled {
    cursor: not-allowed;
  }

  &--shadow {
    box-shadow: var(--button-shadow);
    &:hover {
      box-shadow: var(--button-shadow-hover);
      transform: translate3d(0px, -1px, 0px);
    }
  }

  &--auto {
    width: auto;
    min-width: min-content;
  }

  &--loading {
    pointer-events: none;
  }

  &:not(.fect-button--shadow, .fect-button--disabled, .fect-button--loading):hover,
  &:not(.fect-button--shadow, .fect-button--disabled, .fect-button--loading):focus {
    color: var(--button-hover-color);
    background-color: var(--button-hover-bg-color);
    border-color: var(--button-hover-border-color);
    transform: translate3d(0px, 0px, 0px);
  }

  &__loading {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    z-index: 2;
    background-color: var(--button-bg-color);
    border-radius: var(--button-radius);
  }

  &__drip {
    position: absolute;
    left: 0;
    top: 0;
    right: 0;
    bottom: 0;

    svg {
      position: absolute;
      animation: 350ms ease-in button-expand;
      animation-fill-mode: forwards;
      width: 16px;
      height: 16px;
    }
  }

  &__icon {
    position: absolute;
    top: 50%;
    display: flex;
    justify-content: center;
    align-items: center;
    color: inherit;
    z-index: 1;

    svg {
      background: transparent;
      height: calc(var(--button-height) / 2.35);
      width: calc(var(--button-height) / 2.35);
    }
  }
  &__text {
    position: relative;
    z-index: 1;
    display: inline-flex;
    justify-content: center;
    align-items: center;
    text-align: center;
    line-height: inherit;
    top: -1px;
  }
}
