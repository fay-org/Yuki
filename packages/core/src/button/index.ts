import { withScaleInstall } from '../utils'
import _Button from './button'



export const Button = withScaleInstall(_Button)

export default Button
