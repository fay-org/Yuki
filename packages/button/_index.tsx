import { computed, ref } from 'vue'
import { createNameSpace, useState } from '../utils'
import {
  ButtonTypes,
  LoadingTypes,
  NormalSizes,
} from '../utils/theme/propTypes'
import { CustomCSSProperties } from '../utils/base'

import { props } from './props'
import ButtonLoading from './button-loading'
import { queryHoverColor } from './style'

import './_index.less'
const [createComponent] = createNameSpace('Button')

export default createComponent({
  props,
  emits: ['click'],
  setup(props, { slots, emit }) {
    const buttonRef = ref<HTMLButtonElement>()
    const [drapShow, setDrapShow] = useState<boolean>(false)
    const [drapX, setDrapX] = useState<number>(0)
    const [drapY, setDrapY] = useState<number>(0)

    const clickHandler = (e: MouseEvent) => {
      emit('click', e)
    }

    /**
     * set ghost disabled shadow className
     */
    const setButtonStatus = computed(() => {
      const { ghost, disabled, shadow } = props
      const names = []
      disabled && names.push('is-disabled')
      ghost && names.push('is-ghost')
      shadow && names.push('is-shadow')
      return names.join(' ')
    })

    const setStyle = computed(() => {
      const { type, ghost } = props
      //  return ghost?'a'
      const { bg, border, color } = queryHoverColor(type, ghost)
      const style: CustomCSSProperties = {
        '--button-hover-bg': bg,
        '--button-hover-border': border,
        '--button-hover-color': color,
      }
      return style
    })

    return () => (
      <button
        class={`fect-button 
        fect-button--${props.type} 
        fect-button--${props.size}
        ${setButtonStatus.value}`}
        ref={buttonRef}
        style={setStyle.value}
        onClick={clickHandler}
      >
        {props.loading && <ButtonLoading loadType={props.loadType} />}
        {slots.default?.()}
      </button>
    )
  },
})
