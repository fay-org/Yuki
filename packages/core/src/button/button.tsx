import { computed, ref, defineComponent } from 'vue'
import { useState } from '@fect-ui/vue-hooks'
import { useScale } from '@fect-ui/scale'
import { useTheme } from '../provider/theme-context'
import { createName, createBem, pickContextProps } from '../utils'
import { useButtonGroupContext } from '../button-group/button-group-context'
import { props } from './props'
import ButtonLoading from './button-loading'
import { getButtonColors, getButtonHoverColors, getButtonDripColor } from './style'
import ButtonDrip from './button-drip'

import './index.less'

const name = createName('Button')
const bem = createBem('fect-button')

export default defineComponent({
  name,
  props,
  emits: ['click'],
  setup(props, { slots, emit }) {
    const scale = useScale()
    const { theme } = useTheme()
    const buttonRef = ref<HTMLButtonElement>()
    const [drapShow, setDrapShow] = useState<boolean>(false)
    const [dripX, setDripX] = useState<number>(0)
    const [dripY, setDripY] = useState<number>(0)

    const { context } = useButtonGroupContext()

    const showDrip = computed(() => {
      const { shadow, ghost, effect } = props
      return !shadow && !ghost && effect
    })

    const padding = computed(() => {
      const { SCALES } = scale
      if (props.auto) return [SCALES.pl(1.15), SCALES.pr(1.15)]
      return [SCALES.pl(1.375), SCALES.pr(1.375)]
    })

    const hover = computed(() => getButtonHoverColors(theme.value.palette, filterPropsWithButtonGroup.value))

    const colors = computed(() => getButtonColors(theme.value.palette, filterPropsWithButtonGroup.value))

    const dripColor = computed(() => getButtonDripColor(theme.value.palette, props))

    const setCssVariables = computed(() => {
      const { SCALES } = scale
      const [pl, pr] = padding.value
      const { color, bg, border } = colors.value
      const { bg: hoverBg, border: hoverBorder, color: hoverColor } = hover.value
      const { layout, expressiveness } = theme.value
      return {
        '--button-height': SCALES.height(2.5),
        '--button-min-width': SCALES.width(10.5),
        '--button-radius': layout.radius,
        '--button-font-size': SCALES.font(0.875),
        '--button-shadow': expressiveness.shadowSmall,
        '--button-shadow-hover': expressiveness.shadowMedium,
        '--button-pt': SCALES.pt(0),
        '--button-pr': pr,
        '--button-pb': SCALES.pb(0),
        '--button-pl': pl,
        '--button-mt': SCALES.mt(0),
        '--button-mr': SCALES.mr(0),
        '--button-mb': SCALES.mb(0),
        '--button-ml': SCALES.ml(0),
        '--button-bg-color': bg,
        '--button-border-color': border,
        '--button-color': color,
        '--button-hover-bg-color': hoverBg,
        '--button-hover-border-color': hoverBorder,
        '--button-hover-color': hoverColor
      }
    })

    const clickHandler = (e: MouseEvent) => {
      const { disabled, loading } = filterPropsWithButtonGroup.value
      if (disabled || loading) return
      if (showDrip.value && buttonRef.value) {
        setDrapShow(true)
        const rect = buttonRef.value.getBoundingClientRect()
        setDripX(e.clientX - rect.left)
        setDripY(e.clientY - rect.top)
      }
      emit('click', e)
    }

    const filterPropsWithButtonGroup = computed(() => {
      const { shadow, loading, auto, type, disabled, ghost } = props
      const behavior = pickContextProps({ shadow, loading, auto, type, disabled, ghost }, context)
      if (!context) return { ...behavior, auto }
      return { ...behavior, auto: true }
    })

    const setButtonClasses = computed(() => {
      const state = filterPropsWithButtonGroup.value
      return bem(null, state)
    })

    const dripCompleteHandler = () => {
      setDrapShow(false)
      setDripX(0)
      setDripY(0)
    }

    const renderContext = () => {
      const { icon, default: context } = slots
      if (icon) {
        const offsetStyle = context ? '15%' : '50%'
        const translateStyle = context ? 'translateY(-50%)' : 'translate(-50%, -50%)'
        return (
          <>
            <div class={bem('icon')} style={{ left: offsetStyle, transform: translateStyle }}>
              {icon()}
            </div>
            {context && <div class={bem('text')}>{context()}</div>}
          </>
        )
      }
      return slots.default?.()
    }

    return () => (
      <button
        class={setButtonClasses.value}
        ref={buttonRef}
        style={setCssVariables.value}
        type={props.htmlType}
        onClick={clickHandler}
      >
        {props.loading && <ButtonLoading loadType={props.loadType} color={colors.value.color} />}
        {drapShow.value && (
          <ButtonDrip color={dripColor.value} x={dripX.value} y={dripY.value} onCompleted={dripCompleteHandler} />
        )}
        {renderContext()}
      </button>
    )
  }
})
