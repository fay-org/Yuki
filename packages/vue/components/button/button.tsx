import { computed, ref, defineComponent } from 'vue'
import { useState } from '@fect-ui/vue-hooks'
import { createName, CustomCSSProperties, createBem } from '../utils'
import { props } from './props'
import ButtonLoading from './button-loading'
import { useButtonGroupContext } from '../button-group/button-group-context'
import ButtonDrip from './button-drip'
import { queryHoverColor } from './style'

import './index.less'

const name = createName('Button')

export default defineComponent({
  name,
  props,
  emits: ['click'],
  setup(props, { slots, emit }) {
    const buttonRef = ref<HTMLButtonElement>()
    const [drapShow, setDrapShow] = useState<boolean>(false)
    const [drapX, setDrapX] = useState<number>(0)
    const [drapY, setDrapY] = useState<number>(0)

    const { context } = useButtonGroupContext()

    const clickHandler = (e: MouseEvent) => {
      const { disabled, loading, shadow, ghost, effect } = props
      if (disabled || loading) return
      const showDrip = !shadow && !ghost && effect
      if (showDrip) {
        setDrapShow(true)
        const rect = buttonRef.value!.getBoundingClientRect()!
        setDrapX(e.clientX - rect.left)
        setDrapY(e.clientY - rect.top)
      }
      emit('click', e)
    }

    /**
     * set ghost disabled shadow className
     */
    const setButtonStatus = computed(() => {
      const { ghost, disabled, shadow, auto, loading } = props
      const names = []
      disabled && names.push('is-disabled')
      ghost && names.push('is-ghost')
      shadow && names.push('is-shadow')
      loading && names.push('is-loading')

      if (context && context.props.auto) {
        names.push('is-auto')
      } else {
        auto && names.push('is-auto')
      }

      return names.join(' ')
    })

    const setStyle = computed(() => {
      const { type, ghost } = props
      const { bg, border, color } = queryHoverColor(type, ghost)
      const style: CustomCSSProperties = {
        '--button-hover-bg': bg,
        '--button-hover-border': border,
        '--button-hover-color': color
      }
      return style
    })

    const dripCompleteHandler = () => {
      setDrapShow(false)
      setDrapX(0)
      setDrapY(0)
    }

    const renderContext = () => {
      const { icon } = slots
      const context = slots['default']
      if (icon) {
        const offsetStyle = context ? '15%' : '50%'
        const translateStyle = context ? 'translateY(-50%)' : 'translate(-50%, -50%)'
        return (
          <>
            <div class="fect-button__icon" style={{ left: offsetStyle, transform: translateStyle }}>
              {icon()}
            </div>
            {context && <div class="fect-button__text">{context()}</div>}
          </>
        )
      }
      return slots.default?.()
    }

    const setButtonState = () => {
      let size = ''
      if (context && context.props.size) {
        ;({ size } = context.props)
      } else {
        ;({ size } = props)
      }
      return createBem('fect-button', size)
    }

    return () => (
      <button
        class={`fect-button ${createBem('fect-button', props.type)} ${setButtonState()} ${setButtonStatus.value}`}
        ref={buttonRef}
        style={setStyle.value}
        type={props.htmlType}
        onClick={clickHandler}
      >
        {props.loading && <ButtonLoading loadType={props.loadType} />}
        {drapShow.value && <ButtonDrip x={drapX.value} y={drapY.value} onCompleted={dripCompleteHandler} />}
        {renderContext()}
      </button>
    )
  }
})
