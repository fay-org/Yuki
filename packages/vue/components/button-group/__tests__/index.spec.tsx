import { mount } from '@vue/test-utils'
import { ButtonGroup } from '..'
import { Button } from '../../button'

describe('ButtonGroup', () => {
  it('should render correctly', () => {
    const wrapper = mount({
      render() {
        return (
          <ButtonGroup size="mini">
            <Button>Action</Button>
            <Button>Action</Button>
            <Button>Action</Button>
          </ButtonGroup>
        )
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('should support render vertical', () => {
    const wrapper = mount({
      render() {
        return (
          <ButtonGroup size="mini" vertical>
            <Button>Action</Button>
            <Button>Action</Button>
            <Button>Action</Button>
          </ButtonGroup>
        )
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
