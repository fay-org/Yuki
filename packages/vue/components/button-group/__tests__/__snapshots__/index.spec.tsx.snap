// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup should render correctly 1`] = `
<div class="fect-button__group fect-button__group--horizontal"><button class="fect-button fect-button--default fect-button--mini is-auto" style="--button-hover-bg: var(--button-default-color); --button-hover-border: var(--primary-foreground); --button-hover-color: var(--primary-foreground);" type="button">
    <!---->
    <!---->Action
  </button><button class="fect-button fect-button--default fect-button--mini is-auto" style="--button-hover-bg: var(--button-default-color); --button-hover-border: var(--primary-foreground); --button-hover-color: var(--primary-foreground);" type="button">
    <!---->
    <!---->Action
  </button><button class="fect-button fect-button--default fect-button--mini is-auto" style="--button-hover-bg: var(--button-default-color); --button-hover-border: var(--primary-foreground); --button-hover-color: var(--primary-foreground);" type="button">
    <!---->
    <!---->Action
  </button></div>
`;

exports[`ButtonGroup should support render vertical 1`] = `
<div class="fect-button__group fect-button__group--vertical"><button class="fect-button fect-button--default fect-button--mini is-auto" style="--button-hover-bg: var(--button-default-color); --button-hover-border: var(--primary-foreground); --button-hover-color: var(--primary-foreground);" type="button">
    <!---->
    <!---->Action
  </button><button class="fect-button fect-button--default fect-button--mini is-auto" style="--button-hover-bg: var(--button-default-color); --button-hover-border: var(--primary-foreground); --button-hover-color: var(--primary-foreground);" type="button">
    <!---->
    <!---->Action
  </button><button class="fect-button fect-button--default fect-button--mini is-auto" style="--button-hover-bg: var(--button-default-color); --button-hover-border: var(--primary-foreground); --button-hover-color: var(--primary-foreground);" type="button">
    <!---->
    <!---->Action
  </button></div>
`;
